crates/distribution-types/src/resolved.rs

```diff
@@ -0,0 +1,132 @@
+use std::fmt::Display;
+
+use pep508_rs::PackageName;
+
+use crate::{
+    Dist, DistributionId, DistributionMetadata, Identifier, InstalledDist, Name, ResourceId,
+    VersionOrUrl,
+};
+
+/// A distribution that can be used for resolution and installation.
+///
+/// Either an already-installed distribution or a distribution that can be installed.
+#[derive(Debug, Clone)]
+pub enum ResolvedDist {
+    Installed(InstalledDist),
+    Installable(Dist),
+}
+
+/// A variant of [`ResolvedDist`] with borrowed inner distributions.
+#[derive(Debug, Clone)]
+pub enum ResolvedDistRef<'a> {
+    Installed(&'a InstalledDist),
+    Installable(&'a Dist),
+}
+
+impl ResolvedDist {
+    /// Return true if the distribution is editable.
+    pub fn is_editable(&self) -> bool {
+        match self {
+            Self::Installable(dist) => dist.is_editable(),
+            Self::Installed(dist) => dist.is_editable(),
+        }
+    }
+}
+
+impl ResolvedDistRef<'_> {
+    pub fn as_owned(&self) -> ResolvedDist {
+        match self {
+            Self::Installable(dist) => ResolvedDist::Installable((*dist).clone()),
+            Self::Installed(dist) => ResolvedDist::Installed((*dist).clone()),
+        }
+    }
```

Comment:
Is there a canonical name for this?

Reply:
Perhaps `cloned()`?

Reply:
In the `Cow` impl it's called `into_owned` but `into` usually consumes the value. I'd say `to_owned`: https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

Reply:
Hm I initially did `to_owned` but got confused about how that interacts with the `ToOwned` trait, maybe I'll try to figure that out.

Reply:
e.g.

```
❯ gd
diff --git a/crates/distribution-types/src/resolved.rs b/crates/distribution-types/src/resolved.rs
index fda37bc4..bf7d897d 100644
--- a/crates/distribution-types/src/resolved.rs
+++ b/crates/distribution-types/src/resolved.rs
@@ -33,8 +33,10 @@ impl ResolvedDist {
     }
 }
 
-impl ResolvedDistRef<'_> {
-    pub fn as_owned(&self) -> ResolvedDist {
+impl ToOwned for ResolvedDistRef<'_> {
+    type Owned = ResolvedDist;
+
+    fn to_owned(&self) -> Self::Owned {
         match self {
             Self::Installable(dist) => ResolvedDist::Installable((*dist).clone()),
             Self::Installed(dist) => ResolvedDist::Installed((*dist).clone()),
❯ cc
    Checking distribution-types v0.0.1 (/Users/mz/workspace/uv/crates/distribution-types)
error[E0119]: conflicting implementations of trait `ToOwned` for type `resolved::ResolvedDistRef<'_>`
  --> crates/distribution-types/src/resolved.rs:36:1
   |
36 | impl ToOwned for ResolvedDistRef<'_> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `alloc`:
           - impl<T> ToOwned for T
             where T: Clone;

For more information about this error, try `rustc --explain E0119`.
error: could not compile `distribution-types` (lib) due to 1 previous error
```

Reply:
(just writing `to_owned` without implementing the trait is weird too...)

Reply:
Yeah `to_owned` may not be quite right... I've honestly never implemented the trait so I'm not sure. I do think it should be `to_` something though.

Reply:
Yeah :/ I was pursuing it based on that concept you linked. Unfortunately there isn't a great `to_<...>` subject other than "owned". I'm leaning towards `cloned()`.

Calling in @BurntSushi to tell me what to do.

Reply:
Some interesting context in https://stackoverflow.com/questions/72105604/implement-toowned-for-user-defined-types

Reply:
Yeah it's uncommon to impl `ToOwned` directly. I don't think there's any issue with a concrete `to_owned()` method though. `as_owned()` is definitely not in keeping with convention. 

You could also go the more concrete direction and name this `to_resolved_dist`. That's probably what I would do here I think. But `to_owned()` is also okay IMO.

`cloned()` doesn't feel quite right to me here, since this isn't just a clone, but a conversion.

Reply:
When I did a concrete `to_owned` method it wasn't called — the trait one was called instead. Seems problematic to need to specify the type to lookup the method on?

Reply:
Ahhhhh right okay. Yeah, I'd go with `to_resolved_dist` then personally.

Reply:
I think this might be because the method wasn't public 😬